### PR TITLE
Circuit for opcode SELFBALANCE

### PR DESCRIPTION
**Pull request by scroll-dev**
Implement the SelfbalanceGadget for opcode `SELFBALANCE`.

Specs: https://github.com/appliedzkp/zkevm-specs/pull/113

Will update the bus mapping for SELFBALANCE once https://github.com/appliedzkp/zkevm-circuits/pull/292 is merged in upstream.

### CODE REVIEW

**Comment by ChihChengLiang** [comment]
@scroll-dev Mason Liang Can we do a rebase for this?

File: bus-mapping/src/evm/opcodes.rs [hunk 1]
```diff
@@ -116,7 +117,7 @@ fn fn_gen_associated_ops(opcode_id: &OpcodeId) -> FnGenAssociatedOps {
         // OpcodeId::DIFFICULTY => {},
         // OpcodeId::GASLIMIT => {},
         // OpcodeId::CHAINID => {},
-        // OpcodeId::SELFBALANCE => {},
+        OpcodeId::SELFBALANCE => StackOnlyOpcode::<0>::gen_associated_ops,
```
Comment (ed255) [inline bus-mapping/src/evm/opcodes.rs:120]: Shouldn't there be an CallContextOp and AccountOp READ in this opcode?  If you plan on adding those once #292 is merged, maybe you could set this PR as draft until it's ready?

**Review by ed255 — CHANGES_REQUESTED** [review]

File: zkevm-circuits/src/evm_circuit/execution/selfbalance.rs [hunk 2]
```diff
@@ -0,0 +1,188 @@
...
+
+    fn configure(cb: &mut ConstraintBuilder<F>) -> Self {
+        let callee_address = cb.query_cell();
+        cb.call_context_lookup(
+            None,
+            CallContextFieldTag::CalleeAddress,
+            callee_address.expr(),
+        );
+
+        let self_balance = cb.query_rlc();
```
Comment (ed255) [inline zkevm-circuits/src/evm_circuit/execution/selfbalance.rs:38]: Considering that the `self_balance` value is only used in the RLC encoded form in this gadget, maybe we can avoid using `Word` (which requires 32 cells), and just use a `Cell` for `self_balance`?
I think the current implementation is correct nevertheless.

Comment (z2trillion) [reply zkevm-circuits/src/evm_circuit/execution/selfbalance.rs]: updated the PR to use `Cell` instead.

**Review by ed255 — APPROVED** [review]
LGTM! Thanks, good work!

**Comment by z2trillion** [comment]
@ed255, I've rebased and added the bus mapping.

**Review by ed255 — APPROVED** [review]

**Comment by ed255** [comment]
@z2trillion There seems to be a linting error coming from clippy.

**Comment by z2trillion** [comment]
@ed255 it should be fixed now.

**Pull request merged**